wandb/apis/importers/base.py

```diff
@@ -0,0 +1,312 @@
+import json
+import platform
+from abc import ABC, abstractmethod
+from concurrent.futures import ProcessPoolExecutor, as_completed
+from contextlib import contextmanager
+from typing import Any, Dict, Iterable, List, Optional, Tuple
+
+from tqdm.auto import tqdm
+
+import wandb
+from wandb.proto import wandb_internal_pb2 as pb
+from wandb.proto import wandb_telemetry_pb2 as telem_pb
+from wandb.sdk.interface.interface import file_policy_to_enum
+from wandb.sdk.interface.interface_queue import InterfaceQueue
+from wandb.sdk.internal.sender import SendManager
+
+Name = str
+Path = str
+
+
+def coalesce(*arg: Any) -> Any:
+    """Return the first non-none value in the list of arguments.  Similar to ?? in C#"""
+    return next((a for a in arg if a is not None), None)
+
+
+@contextmanager
+def send_manager(root_dir):
+    sm = SendManager.setup(root_dir, resume=False)
+    try:
+        yield sm
+    finally:
+        # flush any remaining records
+        while sm:
+            data = next(sm)
+            sm.send(data)
+        sm.finish()
+
+
+class Run:
+    def __init__(self) -> None:
+        self.interface = InterfaceQueue()
+        self.run_dir = f"./wandb-importer/{self.run_id()}"
+
+    def run_id(self) -> str:
+        _id = wandb.util.generate_id()
+        wandb.termwarn(f"`run_id` not specified.  Autogenerating id: {_id}")
+        return _id
+
+    def entity(self) -> str:
+        _entity = "unspecified-entity"
+        wandb.termwarn(f"`entity` not specified.  Defaulting to: {_entity}")
+        return _entity
+
+    def project(self) -> str:
+        _project = "unspecified-project"
+        wandb.termwarn(f"`project` not specified.  Defaulting to: {_project}")
+        return _project
+
+    def config(self) -> Dict[str, Any]:
+        return {}
+
+    def summary(self) -> Dict[str, float]:
+        return {}
+
+    def metrics(self) -> List[Dict[str, float]]:
+        """
+        We expect metrics in this shape:
+        [
+            {'metric1': 1, 'metric2': 1, '_step': 0},
+            {'metric1': 2, 'metric2': 4, '_step': 1},
+            {'metric1': 3, 'metric2': 9, '_step': 2},
+            ...
+        ]
+
+        You can also submit metrics in this shape:
+        [
+            {'metric1': 1, '_step': 0},
+            {'metric2': 1, '_step': 0},
+            {'metric1': 2, '_step': 1},
+            {'metric2': 4, '_step': 1},
+            ...
+        ]
+        """
+        return []
+
+    def run_group(self) -> Optional[str]:
+        ...
+
+    def job_type(self) -> Optional[str]:
+        ...
+
+    def display_name(self) -> str:
+        return self.run_id()
+
+    def notes(self) -> Optional[str]:
+        ...
```

Comment:
Wild, never saw this syntax before.  Any particular reason we aren't just using pass?

Reply:
No particular reason -- they are both no-op.  I think it's more common to see `...` in stubs ([example](https://github.com/python/typeshed/blob/main/stdlib/builtins.pyi#L143-L187)).  We can change if you think `pass` is cleaner

Reply:
I'm fine with it, was just curious.  I imagine the SDK folks might have opinions.